src/consensus.rs

```diff
@@ -0,0 +1,264 @@
+        let mut exits = vec![];
+        let mut batch_rxs = vec![];
+        let spawner = Spawner::default();
+
+        for node_ix in 0..n_nodes {
+            let (o, i) = net.consensus_data(node_ix.into());
+            let conf =
+                Config::<NodeId>::new(node_ix.into(), n_nodes.into(), Duration::from_millis(10));
+            let (exit_tx, exit_rx) = oneshot::channel();
+            exits.push(exit_tx);
+            let (batch_tx, batch_rx) = mpsc::unbounded_channel();
+            batch_rxs.push(batch_rx);
+            spawner.spawn(
+                "consensus",
+                Consensus::new(conf, i, o, batch_tx, hashing).run(spawner.clone(), exit_rx),
+            );
+        }
+
+        let mut batches = vec![];
+        for mut rx in batch_rxs.drain(..) {
+            let batch = rx.recv().await.unwrap();
+            assert!(!batch.is_empty());
+            batches.push(batch);
+        }
+
+        // TODO add better checks
+        for node_ix in 1..n_nodes {
+            assert_eq!(batches[0], batches[node_ix]);
+        }
+
+        exits.into_iter().for_each(|tx| {
+            let _ = tx.send(());
+        });
+        spawner.wait().await;
+    }
+
+    #[tokio::test(max_threads = 1)]
+    async fn catches_wrong_control_hash() {
+        init_log();
+        let n_nodes = 4;
+        let spawner = Spawner::default();
+        let node_ix = 0;
+        let (mut tx_in, rx_in) = channel::mpsc::unbounded();
+        let (tx_out, mut rx_out) = channel::mpsc::unbounded();
+        let tx_out = tx_out.sink_map_err(|e| e.into());
+        let conf = Config::<NodeId>::new(node_ix.into(), n_nodes.into(), Duration::from_millis(10));
+        let (exit_tx, exit_rx) = oneshot::channel();
+        let (batch_tx, _batch_rx) = mpsc::unbounded_channel();
+        spawner.spawn(
+            "consensus",
+            Consensus::new(conf, rx_in, tx_out, batch_tx, hashing).run(spawner.clone(), exit_rx),
+        );
+        let mut bad_pu =
+            PreUnit::new_from_parents(1.into(), 0, (vec![None; n_nodes]).into(), hashing);
+        let bad_control_hash = Hash(1111111);
+        assert!(
+            bad_control_hash != bad_pu.control_hash.hash,
+            "Bad control hash cannot be the correct one."
+        );
+        bad_pu.control_hash.hash = bad_control_hash;
+        let bad_hash = Hash(1234567);
+        let bad_unit = Unit::new_from_preunit(bad_pu, bad_hash);
+        let _ = tx_in.send(NotificationIn::NewUnits(vec![bad_unit])).await;
+        loop {
+            let notification = rx_out.next().await.unwrap();
+            debug!("notification {:?}", notification);
+            if let NotificationOut::WrongControlStdHash(h) = notification {
+                assert_eq!(h, bad_hash, "Expected notification for our bad unit.");
+                break;
+            }
+        }
+
+        let _ = exit_tx.send(());
+        spawner.wait().await;
+    }
+
+    // #[tokio::test(max_threads = 1)]
```

Comment:
+1

Reply:
This test has been commented for some time, waiting for this refactor to decide if it can be fixed and adapted to new version or should we just delete it.

Reply:
I would go for deleting as this is from the times when we thought the blocks abstraction will make its way to rush. Now there is Data, so we can trash this.

Reply:
Deleted.